src/main/php/lang/ast/syntax/PHP.class.php

```diff
@@ -1490,14 +1508,10 @@ private function parameters($parse) {
 
       $line= $parse->token->line;
       if ('name' === $parse->token->kind && isset($promotion[$parse->token->value])) {
-        $promote= $parse->token->value;
-        $parse->forward();
-
-        // It would be better to use an array for promote, but this way we keep BC
-        if ('readonly' === $parse->token->value) {
-          $promote.= ' readonly';
-          $parse->forward();
-        }
+        $promote= [];
```

Comment:
Changing `Parameter::$promote` from `?string` to `?array<string>` requires the following patch to the XP Compiler:

```diff
diff --git a/src/main/php/lang/ast/emit/PHP.class.php b/src/main/php/lang/ast/emit/PHP.class.php
index d83c5cc..b59100d 100755
--- a/src/main/php/lang/ast/emit/PHP.class.php
+++ b/src/main/php/lang/ast/emit/PHP.class.php
@@ -718,7 +718,11 @@ abstract class PHP extends Emitter {
     }
 
     foreach ($promoted as $param) {
-      $this->emitProperty($result, new Property(explode(' ', $param->promote), $param->name, $param->type));
+      $this->emitProperty($result, new Property(
+        is_array($param->promote) ? $param->promote : explode(' ', $param->promote),
+        $param->name,
+        $param->type
+      ));
     }
 
     $result->locals= $locals;
diff --git a/src/main/php/lang/ast/emit/ReadonlyClasses.class.php b/src/main/php/lang/ast/emit/ReadonlyClasses.class.php
index 715c9b8..465f94f 100755
--- a/src/main/php/lang/ast/emit/ReadonlyClasses.class.php
+++ b/src/main/php/lang/ast/emit/ReadonlyClasses.class.php
@@ -21,7 +21,7 @@ trait ReadonlyClasses {
           $member->modifiers[]= 'readonly';
         } else if ($member->is('method')) {
           foreach ($member->signature->parameters as $param) {
-            $param->promote && $param->promote.= ' readonly';
+            $param->promote && $param->promote[]= ' readonly';
           }
         }
       }
```

Reply:
Change committed in xp-framework/compiler@dbc916c

Reply:
https://github.com/xp-framework/compiler/releases/tag/v9.1.1